### ISSUE TITLE
Multiple child-CryptoRepoFiles with same parent + localName

### ISSUE THREAD

**Issue by nlmarco** [subshare/subshare#50]
Encountered the following error:

```
2017-05-14 23:15:37,495 [RepoSyncDaemonThread_0_6] ERROR c.c.c.c.r.s.RepoSyncDaemonImpl$WrapperRunnable - run: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.reflect.UndeclaredThrowableException
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.reflect.UndeclaredThrowableException
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.waitForAndCheckLocalSyncFuture(RepoToRepoSync.java:188) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.waitForAndCheckLocalSyncFutureIfExists(RepoToRepoSync.java:179) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:226) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.syncDown(RepoToRepoSync.java:174) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:137) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at org.subshare.core.repo.sync.SsRepoToRepoSync.sync(SsRepoToRepoSync.java:65) ~[org.subshare.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncRunner.run(RepoSyncRunner.java:51) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncDaemonImpl$WrapperRunnable.run(RepoSyncDaemonImpl.java:123) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_131]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_131]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
Caused by: java.util.concurrent.ExecutionException: java.lang.reflect.UndeclaredThrowableException
        at java.util.concurrent.FutureTask.report(FutureTask.java:122) [na:1.8.0_131]
        at java.util.concurrent.FutureTask.get(FutureTask.java:192) [na:1.8.0_131]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.waitForAndCheckLocalSyncFuture(RepoToRepoSync.java:184) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        ... 12 common frames omitted
Caused by: java.lang.reflect.UndeclaredThrowableException: null
        at com.sun.proxy.$Proxy63.localSync(Unknown Source) ~[na:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync$1.call(RepoToRepoSync.java:131) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync$1.call(RepoToRepoSync.java:127) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        ... 4 common frames omitted
Caused by: java.lang.reflect.InvocationTargetException: null
        at sun.reflect.GeneratedMethodAccessor915.invoke(Unknown Source) ~[na:na]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_131]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_131]
        at co.codewizards.cloudstore.local.LocalRepoManagerInvocationHandler.invoke(LocalRepoManagerInvocationHandler.java:65) ~[co.codewizards.cloudstore.local-0.10.0.jar:na]
        ... 7 common frames omitted
Caused by: javax.jdo.JDOUserException: The query returned more than one instance BUT either unique is set to true or only aggregates are to be returned, so should have returned one result maximum
        at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:639) ~[datanucleus-api-jdo-4.0.3.jar:na]
        at org.datanucleus.api.jdo.JDOQuery.execute(JDOQuery.java:277) ~[datanucleus-api-jdo-4.0.3.jar:na]
        at org.subshare.local.persistence.CryptoRepoFileDao.getChildCryptoRepoFile(CryptoRepoFileDao.java:123) ~[org.subshare.local-0.10.0.jar:na]
        at org.subshare.local.CryptreeNode.getCryptoRepoFile(CryptreeNode.java:242) ~[org.subshare.local-0.10.0.jar:na]
        at org.subshare.local.CryptreeImpl.preDelete(CryptreeImpl.java:2369) ~[org.subshare.local-0.10.0.jar:na]
        at org.subshare.local.SsLocalRepoSync.deleteRepoFileWithAllChildrenRecursively(SsLocalRepoSync.java:309) ~[org.subshare.local-0.10.0.jar:na]
        at co.codewizards.cloudstore.local.LocalRepoSync.deleteRepoFile(LocalRepoSync.java:374) ~[co.codewizards.cloudstore.local-0.10.0.jar:na]
        at co.codewizards.cloudstore.local.LocalRepoSync.deleteRepoFile(LocalRepoSync.java:358) ~[co.codewizards.cloudstore.local-0.10.0.jar:na]
        at co.codewizards.cloudstore.local.LocalRepoSync.sync(LocalRepoSync.java:194) ~[co.codewizards.cloudstore.local-0.10.0.jar:na]
        at org.subshare.local.SsLocalRepoSync.sync(SsLocalRepoSync.java:121) ~[org.subshare.local-0.10.0.jar:na]
        at co.codewizards.cloudstore.local.LocalRepoSync.sync(LocalRepoSync.java:185) ~[co.codewizards.cloudstore.local-0.10.0.jar:na]
        at org.subshare.local.SsLocalRepoSync.sync(SsLocalRepoSync.java:121) ~[org.subshare.local-0.10.0.jar:na]
        at co.codewizards.cloudstore.local.LocalRepoSync.sync(LocalRepoSync.java:78) ~[co.codewizards.cloudstore.local-0.10.0.jar:na]
        at org.subshare.local.SsLocalRepoSync.sync(SsLocalRepoSync.java:68) ~[org.subshare.local-0.10.0.jar:na]
        at co.codewizards.cloudstore.local.LocalRepoManagerImpl.localSync(LocalRepoManagerImpl.java:839) ~[co.codewizards.cloudstore.local-0.10.0.jar:na]
        ... 11 common frames omitted
Caused by: org.datanucleus.store.query.QueryNotUniqueException: The query returned more than one instance BUT either unique is set to true or only aggregates are to be returned, so should have returned one result maximum
        at org.datanucleus.store.query.Query.executeQuery(Query.java:1875) ~[datanucleus-core-4.0.3.jar:na]
        at org.datanucleus.store.query.Query.executeWithArray(Query.java:1729) ~[datanucleus-core-4.0.3.jar:na]
        at org.datanucleus.api.jdo.JDOQuery.execute(JDOQuery.java:268) ~[datanucleus-api-jdo-4.0.3.jar:na]
        ... 24 common frames omitted
```

**Comment by nlmarco** [comment]
This definitely is an illegal state! IMHO, this can only happen in case of a collision and then `DuplicateCryptoRepoFileHandler` deduplicates it.

Hmmm... Thinking further about it: Maybe, there are permission-related situations, too, in which this might happen. For example, if user A created and deleted a file which user B is not allowed to see. User B then creates the same file -- and user A who is allowed to read both his own old and B's new file, has this problem.

However, this happened to me without any permissions problems, because I definitely was the only one (on 2 different computers) to create this file (and delete it again).

Yesterday, I changed the query to be non-UNIQUE and instead throw an exception in the Java code. This is the new error message:

`2017-05-16 18:18:02,199 [pool-4-thread-1] ERROR o.s.l.p.CryptoRepoFileDao - getChildCryptoRepoFile: Expected 0 or 1, but found multiple child-CryptoRepoFiles! parentCryptoRepoFileId=mnC_qsDBrVk9McVrpPiNIQ, parentLocalName=IT, localName=marco.kdb.lock, childCryptoRepoFiles=[CryptoRepoFile[3909]{cryptoRepoFileId=UIr6UHr3b5V0hddv1LgWVA, localName=marco.kdb.lock, deleted=Sun Jul 17 19:04:31 CEST 2016}, CryptoRepoFile[3910]{cryptoRepoFileId=bxl-yTj7oDNwdG09XTAcUQ, localName=marco.kdb.lock, deleted=Mon Jul 18 17:19:37 CEST 2016}, CryptoRepoFile[3918]{cryptoRepoFileId=5WfBzdHaZf0m1tWzvPhIKg, localName=marco.kdb.lock, deleted=Sun Jul 24 05:23:24 CEST 2016}, CryptoRepoFile[3928]{cryptoRepoFileId=FpxZTkUrwFhsWkUhRpukfg, localName=marco.kdb.lock, deleted=Tue Jul 26 07:20:27 CEST 2016}, CryptoRepoFile[3991]{cryptoRepoFileId=qsLB-XQfQ2hHhqpZsNcgrw, localName=marco.kdb.lock, deleted=Tue Aug 09 19:06:13 CEST 2016}, CryptoRepoFile[3996]{cryptoRepoFileId=xZ9J74Vx4iTO2hA6RNyaXQ, localName=marco.kdb.lock, deleted=Mon Aug 15 17:12:49 CEST 2016}]`

**Comment by nlmarco** [comment]
A file that's deleted and then recreated causes the `CryptoRepoFile` to be recycled. `CryptreeNode.getCryptoRepoFileOrCreate(...)` does this (it sets the `deleted` property to `null`). But I just saw that the code looking for a `CryptoRepoFile` by its `parent` and its `localName` was added 2016-09-18 -- **after** the deleted-timestamps of the entities logged in the error above.

This thus seems to be a legacy problem and I fix it -- at least for now -- with a workaround.

But we definitely should think about this even further -- maybe it really may happen legally when a non-privileged user does not see a file that already exists and creates it with the same name. IMHO, the deduplication logic only is triggered, if both files still exist and thus cause a collision. But if the privileged user already deleted it, then there is likely no collision and thus no deduplication.
